autogen/browser_utils.py

```diff
@@ -280,3 +313,741 @@ def _fetch_page(self, url: str) -> None:
+
+    Raises:
+        ImportError: If selenium package is not installed, it raises an ImportError with a message suggesting to install it using pip.
+    """
+
+    # Load the arguments from kwargs
+    browser = kwargs.get("browser", "edge")
+    download_dir = kwargs.get("download_dir", tempfile.gettempdir())
+    if not download_dir:
+        download_dir = tempfile.gettempdir()
+
+    browser_res = kwargs.get("resolution", (1920, 5200))
+
+    def get_headless_options(download_dir, options):
+        options.headless = True
+        options.add_argument("--headless")
+        options.add_argument(f"--window-size={browser_res[0]},{browser_res[1]}")
+        options.add_argument("--downloadsEnabled")
+        if download_dir:
+            options.set_preference("download.default_directory", download_dir)
+        return options
+
+    if browser.lower() == "edge":
+        options = EdgeOptions()
+        options.use_chromium = True  # Ensure we're using the Chromium-based version of Edge
+        options.headless = True
+        options.add_argument("--headless")
+        options.add_argument(f"--window-size={browser_res[0]},{browser_res[1]}")
+        options.add_argument("--downloadsEnabled")
+
+        prefs = {
+            "download.default_directory": download_dir,
+            "download.prompt_for_download": False,  # Disable download prompt
+            "download.directory_upgrade": True,  # Enable directory upgrade
+            "safebrowsing.enabled": True,  # Enable safe browsing
+        }
+        options.add_experimental_option("prefs", prefs)
+        # Instantiate the EdgeService object
+        edge_service = EdgeService()
+        # Instantiate the Edge WebDriver with the configured options
+        driver = webdriver.Edge(options=options, service=edge_service)
+
+    elif browser.lower() == "firefox":
+        # Instantiate the Firefox Profile to specify options
+        profile = FirefoxProfile()
+        profile.set_preference("browser.download.folderList", 2)  # Custom location
+        profile.set_preference("browser.download.dir", download_dir)
+        profile.set_preference("browser.download.useDownloadDir", True)
+        profile.set_preference("browser.helperApps.neverAsk.saveToDisk", "application/pdf")  # MIME type
+        # profile.set_preference("pdfjs.disabled", True) # Disable PDF viewer
+        profile.set_preference("javascript.enabled", False)
+        # profile.set_preference("browser.startup.homepage", "https://microsoft.com")
+        profile.update_preferences()
+        options = FirefoxOptions()
+        options.profile = profile
+        options.set_capability("se:downloadsEnabled", True)
+
+        # Instantiate the Firefox WebDriver with the configured options
+        driver = webdriver.Firefox(
+            options=get_headless_options(download_dir, options)
+        )  # , service_log_path=f'{tempfile.tempdir}/geckodriver.log')
```

Comment:
I can delete this extra comment, and a few others above it.  They are remnants from ensuring coverage during the graphical websurfer notebook.